### ISSUE TITLE
Require server authentication in TLS clients

### ISSUE THREAD

**Issue by gilles-peskine-arm** [Mbed-TLS/mbedtls#7080]
It is currently possible to configure authentication as required, optional or skipped, both for TLS clients and for TLS servers. This makes sense in servers: some applications want only authenticated clients, some applications don't care about client authentication, and some applications allow arbitrary clients but grant authenticated clients extra privileges. But it doesn't make sense for TLS servers. In common settings, TLS offers no security if the server is not authenticated. (Using TLS without authenticating the server only has any security advantage if adversaries can eavesdrop, but not inject or modify traffic.)

Accepting unauthenticated servers is a high-risk decision, so it would make sense to make it less easy than just one easy-to-call configuration function. Currently, the documentation of `mbedtls_ssl_conf_authmode` states that “`MBEDTLS_SSL_VERIFY_REQUIRED` is the recommended mode” and explains how to do server authentication with `MBEDTLS_SSL_VERIFY_OPTIONAL`, but does not highlight all the risks.

If an application wants to allow unauthenticated clients, a workaround exists: call `mbedtls_ssl_conf_ca_cb` and write a custom function that returns true without authenticating the server.

In TLS 1.3, as of Mbed TLS 3.3.0, [server authentication is always performed regardless of the authmode setting](https://github.com/Mbed-TLS/mbedtls/issues/7075).

Proposal: make `mbedtls_ssl_conf_authmode` a no-op on clients. Change the documentation accordingly.

**Comment by mpg** [comment]
> In common settings, TLS offers no security if the server is not authenticated. (Using TLS without authenticating the server only has any security advantage if adversaries can eavesdrop, but not inject or modify traffic.)

I don't know if that's still the case, but a few years ago there were important use cases (SMTP) were authenticated TLS was not always possible and having the option to fall back gradually to unauthenticated TLS was seen by many as a better than falling back all the way to clear text. More generally, while clearly we want to nudge best users towards the most secure practices, we may want to avoid presenting them with all-or-nothing king of choices about security, because we don't want them to pick "nothing" because "some security, even if it's not perfect" was no longer an option.

That said, I agree with the general idea here, just wanted to emphasize that it's important to think carefully about workarounds and document them. We might even want to have a simple example program about that. And take this opportunity to encourage users to be as fine-grained as possible in what they ignore. For example, if they can't fully verify certs because they don't have a valid clock, they should use a verify callback that will change the fail to a pass if the only flags were about time validity, but still reject those with invalid signatures, and it would be good for an example program to show that kind of thing.

**Comment by mpg** [comment]
I'll also note that using a callback to force-accept a certificate is not equivalent to `MBEDTLS_SSL_VERIFY_NONE`: with `VERIFY_NONE` we don't even parse the presented certificate, so the connection will proceed even if the peers presents a certificate that we would fail to parse (for example because it includes an extension that we don't support and that's marked as critical). I don't think that can be achieved with any of the existing callbacks.

I don't know if that's a use case we want to keep supporting, I'm just raising the question.

**Comment by gilles-peskine-arm** [comment]
Note that our sample code relies on disabling server authentication (and as a consequence [broke with TLS 1.3](https://github.com/Mbed-TLS/mbedtls/issues/9223)) — see https://github.com/Mbed-TLS/mbedtls/issues/9079. So this task is more work than the obvious:

* Remove `MBEDTLS_SSL_VERIFY_OPTIONAL` and `MBEDTLS_SSL_VERIFY_NONE` on clients.
* This probably means we want to tweak the functions themselves, since now there's no longer an `authmode` to change on clients, only on servers.
* And we need to resolve https://github.com/Mbed-TLS/mbedtls/issues/9079, and possibly also some test code might rely on skipping authentication (TBD).

**Comment by irwir** [comment]
> I don't know if that's still the case, but a few years ago there were important use cases (SMTP) were authenticated TLS was not always possible and having the option to fall back gradually to unauthenticated TLS was seen by many as a better than falling back all the way to clear text.

Even worse than `clear text` - it might be impossible to connect at all.
Another case would be use of self-signed certificates for remote connections in browser. Using private key in someone else's device is out of question. But user could view the received certificate and allow exception.

**Comment by CodingMarkus** [comment]
Our application has always verified server certificates before allowing a TLS handshake to continue but it must use an external verification system. That's why we set `MBEDTLS_SSL_VERIFY_NONE` even as a client, so we can perform our own verification. If you remove that capabilities from mbedTLS, we must look for a new TLS library.

We cannot use `mbedtls_ssl_conf_ca_cb`, as we have no access to any CAs. The CAs are in a closed system whose only API is to send it a certificate together with some meta info (like the domain/IP address you are talking to) and the result will then either be `fine, you can trust that cert` or `you cannot trust that cert, reason: ...`. 

We need access to the cert and we need a way to tell mbedTLS if that cert is trusted or not. If that is done via a new callback, fine. But going by your current plans, I have to stop the mbedTLS 3.x migration right now, rollback to 2.x and start looking for an alternative library which continues to allow us performing validation ourselves even when using TLS 1.3

**Comment by gilles-peskine-arm** [comment]
@CodingMarkus In Mbed TLS 3.6.1, we will [align the TLS 1.3 behavior with TLS 1.2](https://github.com/Mbed-TLS/mbedtls/issues/9018), so you will be able to continue using `MBEDTLS_SSL_VERIFY_NONE`.

In Mbed TLS 4.0, we will make server authentication mandatory. I do note that the current certificate verification callback is insufficient. Thanks for sharing this use case. I'm adding this as a requirement here.

**Comment by CodingMarkus** [comment]
> I do note that the current certificate verification callback is insufficient. 

Would you like me to create an enhancement request for that?

Currently we use `mbedtls_ssl_conf_verify()` and if the certificate does validate, we continue with the handshake which we perform single step by step, otherwise we do return a meaningful error (as good as we can) but only to allow internal clean up, because we will never call `mbedtls_ssl_handshake_step()` again, never allow any traffic from/to the TCP socket and close the TCP socket with RST, which may not be *nice* but is probably the safest thing we can do to prevent any data flowing after the server cert validation has failed.

**Comment by gilles-peskine-arm** [comment]
> Would you like me to create an enhancement request for that?

Yes, please, that would be appreciated! You know your use case best.

> close the TCP socket with RST, which may not be nice but is probably the safest thing we can do to prevent any data flowing after the server cert validation has failed.

It certainly is the safest. I think that as long as no `mbedtls_ssl_xxx` function has returned an error code, you can call `mbedtls_ssl_send_alert_message()` to send an alert. Handshake handling functions do that, so it should be safe to send an alert yourself as long as the SSL context is not in a failure state. But don't call any SSL function, including `mbedtls_ssl_send_alert_message()`, if an SSL function has returned an error (other than the continuable errors: `WANT_READ`, `WANT_WRITE`, `ASYNC_IN_PROGRESS`).

**Comment by ontera** [comment]
> If an application wants to allow unauthenticated clients, a workaround exists: call `mbedtls_ssl_conf_ca_cb` and write a custom function that returns true without authenticating the server.

Sorry to be jumping into this so late, but should this read:

     If an application wants to allow unauthenticated **_servers_**, a workaround exists...

Asking because I have a (consulting, not TLS) client who read this and was confused by it, and asked me about it.  

Thank you!
